java/vector/src/main/java/org/apache/arrow/vector/NullableTimeStampMilliVector.java

```diff
@@ -0,0 +1,301 @@
+/*******************************************************************************
+
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+package org.apache.arrow.vector;
+
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.vector.complex.impl.TimeStampMilliReaderImpl;
+import org.apache.arrow.vector.complex.reader.FieldReader;
+import org.apache.arrow.vector.holders.TimeStampMilliHolder;
+import org.apache.arrow.vector.holders.NullableTimeStampMilliHolder;
+import org.apache.arrow.vector.types.Types;
+import org.apache.arrow.vector.types.pojo.FieldType;
+import org.apache.arrow.vector.util.TransferPair;
+import org.joda.time.LocalDateTime;
+
+/**
+ * NullableTimeStampMilliVector implements a fixed width vector (8 bytes) of
+ * timestamp values which could be null. A validity buffer (bit vector) is
+ * maintained to track which elements in the vector are null.
+ */
+public class NullableTimeStampMilliVector extends BaseNullableFixedWidthVector {
```

Comment:
I suppose we want to consolidate all Timestamp vectors into one class later?

Reply:
Not sure what that buys. Definitely seems like something we should address afterwards (but before a release--better not to break things twice)

Reply:
Yeah, we can address this later as long as it's before release.

Reply:
In C++ we have a single `arrow::TimestampArray` but applications will generally branch based on the unit metadata and look at the int64 values. I can't really say what's the better option for Java

Reply:
In general, in Java we want to avoid having extra branches at the cell level. For long I agree all could be the same. So maybe having a TimestampVector that takes a constructor provided unit type and then have subclassed vectors that hold the old per unit hierarchy could give a good combination. Both would have the get and set with int64/long. If you want to avoid the getValue() conditional, you go with the sub-hiearchy. If you want a generic interface for getValue(), use the base TimestampVector which does an internal conditional on how to interpret the value. Keeps @icexelloss happy but also allows Dremio folk to use the specific interfaces in runtime generated code.

Reply:
Makes sense. I think this is a Java-specific design issue then related to the JIT -- in C/C++ we have to branch earlier than the cell level into vectorized branch-free code paths so having a single container makes things simpler (since you can dispatch to code that handles "some kind of timestamp").

Reply:
Is thing something we should do as part of the ongoing refactor patch? Earlier @jacques-n mentioned doing it afterwards probably once the new infrastructure is stable? I am fine either way. Just confirming..